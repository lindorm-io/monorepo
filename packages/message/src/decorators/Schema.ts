import { Constructor } from "@lindorm/types";
import z, { ZodObject, ZodRawShape } from "zod";
import { IMessage } from "../interfaces";
import { globalMessageMetadata } from "../utils";

export function Schema<T extends ZodObject<ZodRawShape>>(
  schema: T,
): <C extends Constructor<IMessage & z.infer<T>>>(ctor: C) => void {
  return function <C extends Constructor<IMessage & z.infer<T>>>(target: C) {
    globalMessageMetadata.addSchema({
      target,
      schema,
    });
  };
}
