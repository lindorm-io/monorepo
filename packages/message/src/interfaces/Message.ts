export interface IMessage {
  [key: string]: any;
}
