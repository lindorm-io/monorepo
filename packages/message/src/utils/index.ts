export * from "./default-create-message";
export * from "./default-generate-message";
export * from "./default-validate-message";
export * from "./get-topic-name";
export * from "./global";
