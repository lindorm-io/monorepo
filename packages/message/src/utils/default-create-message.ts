import { isFunction } from "@lindorm/is";
import { Constructor, DeepPartial } from "@lindorm/types";
import { IMessage } from "../interfaces";
import { globalMessageMetadata } from "./global";
import { parseField } from "./private";

export const defaultCreateMessage = <
  E extends IMessage,
  O extends DeepPartial<E> = DeepPartial<E>,
>(
  Message: Constructor<E>,
  options: O | E = {} as O,
): E => {
  const metadata = globalMessageMetadata.get(Message);
  const message = new Message();

  for (const field of metadata.fields) {
    (message as any)[field.key] = parseField(field, message, options);

    if ((message as any)[field.key]) continue;

    (message as any)[field.key] = isFunction(field.fallback)
      ? field.fallback()
      : field.fallback;
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnCreate");

  for (const hook of hooks) {
    hook.callback(message);
  }

  return message;
};
