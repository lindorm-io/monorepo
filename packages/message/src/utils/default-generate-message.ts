import { Constructor, DeepPartial } from "@lindorm/types";
import { randomBytes, randomInt, randomUUID } from "crypto";
import { IMessage } from "../interfaces";
import { MessageMetadata, MetaGenerated } from "../types";
import { globalMessageMetadata } from "./global";

const generator = (config: Omit<MetaGenerated, "target">): any => {
  switch (config.strategy) {
    case "date":
      return new Date();

    case "float":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return Math.random() * (config.max - config.min) + config.min;

    case "integer":
      config.min = config.min ?? 0;
      config.max = config.max ?? 999999;
      return randomInt(config.min, config.max);

    case "string":
      config.length = config.length ?? 32;
      return randomBytes(config.length).toString("base64url");

    case "uuid":
      return randomUUID();

    default:
      return null;
  }
};

const validator = (
  config: Omit<MetaGenerated, "target">,
  metadata: MessageMetadata,
): void => {
  const field = metadata.fields.find((item) => item.key === config.key);

  if (!field) {
    throw new Error(`Field not found for key: ${config.key}`);
  }

  if (!field.type) return;

  switch (config.strategy) {
    case "date":
      if (field.type === "date") break;
      throw new Error(`Invalid field type for date generation: ${field.type}`);

    case "float":
      if (field.type === "float") break;
      throw new Error(`Invalid field type for int generation: ${field.type}`);

    case "integer":
      if (field.type === "integer") break;
      throw new Error(`Invalid field type for int generation: ${field.type}`);

    case "string":
      if (field.type === "string") break;
      throw new Error(`Invalid field type for string generation: ${field.type}`);

    case "uuid":
      if (field.type === "uuid") break;
      throw new Error(`Invalid field type for uuid generation: ${field.type}`);

    default:
      break;
  }
};

export const defaultGenerateMessage = <M extends IMessage>(
  target: Constructor<M>,
  message: DeepPartial<M>,
): M => {
  const metadata = globalMessageMetadata.get(target);

  for (const config of metadata.generated) {
    if (message[config.key]) continue;

    validator(config, metadata);

    (message as any)[config.key] = generator(config);
  }

  return message as M;
};
