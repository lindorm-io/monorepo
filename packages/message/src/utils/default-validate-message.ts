import { Constructor } from "@lindorm/types";
import { z, ZodArray, ZodNumber, ZodRawShape, ZodString, ZodType } from "zod";
import { IMessage } from "../interfaces";
import { MetaField } from "../types";
import { globalMessageMetadata } from "./global";

const getValidator = (field: Omit<MetaField, "target">): ZodType<any> | undefined => {
  switch (field.type) {
    case "array":
      return z.array(z.any());

    case "bigint":
      return z.bigint();

    case "boolean":
      return z.boolean();

    case "date":
      return z.date();

    case "email":
      return z.string().email();

    case "enum":
      return z.nativeEnum(field.enum);

    case "float":
      return z.number();

    case "integer":
      return z.number().int();

    case "object":
      return z.object({}).passthrough();

    case "string":
      return z.string();

    case "url":
      return z.string().url();

    case "uuid":
      return z.string().uuid();

    default:
      return;
  }
};

export const defaultValidateMessage = <M extends IMessage>(
  target: Constructor<M>,
  message: M,
): void => {
  const metadata = globalMessageMetadata.get(target);
  const validators: ZodRawShape = {};

  for (const field of metadata.fields) {
    let validator = getValidator(field);

    if (!validator) continue;

    if (field.type && ["array", "number", "string"].includes(field.type) && field.min) {
      validator = (validator as ZodArray<any> | ZodNumber | ZodString).min(field.min);
    }

    if (field.type && ["array", "number", "string"].includes(field.type) && field.max) {
      validator = (validator as ZodArray<any> | ZodNumber | ZodString).max(field.max);
    }

    if (field.nullable) {
      validator = validator.nullable();
    }

    if (field.optional) {
      validator = validator.optional();
    }

    validators[field.key] = validator;
  }

  z.object(validators).passthrough().parse(message);

  if (metadata.schema) {
    metadata.schema.parse(message);
  }

  const hooks = metadata.hooks.filter((h) => h.decorator === "OnValidate");

  for (const hook of hooks) {
    hook.callback(message);
  }
};
