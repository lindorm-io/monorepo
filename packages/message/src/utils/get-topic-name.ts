import { kebabCase } from "@lindorm/case";
import { Constructor } from "@lindorm/types";
import { IMessage } from "../interfaces";
import { globalMessageMetadata } from "./global";

type Options = {
  namespace?: string | null;
  topic?: string | null;
};

export const getTopicName = <M extends IMessage>(
  MessageConstructor: Constructor<M>,
  message: M,
  options: Options,
): string => {
  const metadata = globalMessageMetadata.get(MessageConstructor);

  if (metadata.topic?.callback) {
    return metadata.topic.callback(message);
  }

  if (metadata.message.topic) return metadata.message.topic;
  if (options.topic) return options.topic;

  const namespace = metadata.message.namespace || options.namespace;
  const messageName = metadata.message.name || MessageConstructor.name;

  if (namespace === "system") {
    throw new Error("The 'system' namespace is reserved for internal use");
  }

  const n = namespace ? `${kebabCase(namespace)}.` : "";
  const e = kebabCase(messageName);

  const name = `${n}${e}`;

  if (name.length > 120) {
    throw new Error(`Topic name exceeds 120 characters: ${name}`);
  }

  return name;
};
