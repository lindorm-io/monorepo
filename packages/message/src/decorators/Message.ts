import { MessageDecoratorOptions } from "../types";
import { globalMessageMetadata } from "../utils";

export function Message(options: MessageDecoratorOptions = {}): ClassDecorator {
  return function (target) {
    globalMessageMetadata.addMessage({
      target,
      decorator: "Message",
      name: options?.name || target.name,
      namespace: options?.namespace || null,
      topic: options?.topic || null,
    });
  };
}
