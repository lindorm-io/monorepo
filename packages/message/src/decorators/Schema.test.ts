import { z } from "zod";
import { globalMessageMetadata } from "../utils";
import { Message } from "./Message";
import { Schema } from "./Schema";

describe("Schema Decorator", () => {
  test("should add metadata", () => {
    @Message()
    @Schema(z.object({ name: z.string() }))
    class SchemaDecoratorMessage {}

    expect(globalMessageMetadata.get(SchemaDecoratorMessage)).toMatchSnapshot();
  });
});
