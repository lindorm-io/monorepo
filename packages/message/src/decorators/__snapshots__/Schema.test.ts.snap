// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema Decorator should add metadata 1`] = `
{
  "fields": [
    {
      "decorator": "Field",
      "enum": null,
      "fallback": null,
      "key": "name",
      "max": null,
      "min": null,
      "nullable": false,
      "optional": false,
      "schema": null,
      "type": null,
    },
  ],
  "generated": [],
  "hooks": [],
  "message": {
    "decorator": "Message",
    "name": "SchemaDecoratorMessage",
    "namespace": null,
    "target": [Function],
    "topic": null,
  },
  "priority": undefined,
  "schema": ZodObject {
    "_cached": null,
    "_def": {
      "catchall": ZodNever {
        "_def": {
          "typeName": "ZodNever",
        },
        "and": [Function],
        "array": [Function],
        "brand": [Function],
        "catch": [Function],
        "default": [Function],
        "describe": [Function],
        "isNullable": [Function],
        "isOptional": [Function],
        "nullable": [Function],
        "nullish": [Function],
        "optional": [Function],
        "or": [Function],
        "parse": [Function],
        "parseAsync": [Function],
        "pipe": [Function],
        "promise": [Function],
        "readonly": [Function],
        "refine": [Function],
        "refinement": [Function],
        "safeParse": [Function],
        "safeParseAsync": [Function],
        "spa": [Function],
        "superRefine": [Function],
        "transform": [Function],
        "~standard": {
          "validate": [Function],
          "vendor": "zod",
          "version": 1,
        },
      },
      "shape": [Function],
      "typeName": "ZodObject",
      "unknownKeys": "strip",
    },
    "and": [Function],
    "array": [Function],
    "augment": [Function],
    "brand": [Function],
    "catch": [Function],
    "default": [Function],
    "describe": [Function],
    "isNullable": [Function],
    "isOptional": [Function],
    "nonstrict": [Function],
    "nullable": [Function],
    "nullish": [Function],
    "optional": [Function],
    "or": [Function],
    "parse": [Function],
    "parseAsync": [Function],
    "pipe": [Function],
    "promise": [Function],
    "readonly": [Function],
    "refine": [Function],
    "refinement": [Function],
    "safeParse": [Function],
    "safeParseAsync": [Function],
    "spa": [Function],
    "superRefine": [Function],
    "transform": [Function],
    "~standard": {
      "validate": [Function],
      "vendor": "zod",
      "version": 1,
    },
  },
  "topic": undefined,
}
`;
